fiftyone/operators/cache/utils.py

```
@@ -0,0 +1,239 @@
+    key_string = json.dumps(structured_key, sort_keys=True)
+    return hashlib.sha256(key_string.encode("utf-8")).hexdigest()
+
+
+def _get_cache_key_list(ctx_index, args, kwargs, key_fn):
+    if key_fn:
+        try:
+            cache_key_list = key_fn(*args, **kwargs)
+            if not isinstance(cache_key_list, list):
+                raise ValueError("Custom key function must return a list.")
+        except Exception as e:
+            raise ValueError(f"Failed to create custom cache key: {e}")
```

Ruff (0.8.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)